util/deephash/deephash.go

```diff
@@ -164,26 +164,144 @@ func (h *hasher) hashUint64(i uint64) {
 
 var uint8Type = reflect.TypeOf(byte(0))
 
-func (h *hasher) hashValue(v reflect.Value) {
-	if !v.IsValid() {
-		return
+// typeInfo describes properties of a type.
+type typeInfo struct {
+	rtype       reflect.Type
+	isRecursive bool
+
+	// eleTypeInfo is the element type's typeInfo.
+	// It's set when rtype is of Kind Ptr, Slice, Array, Map.
+	eleTypeInfo *typeInfo
+
+	// keyTypeInfo is the map key type's typeInfo.
+	// It's set when rtype is of Kind Map.
+	keyTypeInfo *typeInfo
+}
+
+var typeInfoMap sync.Map           // map[reflect.Type]*typeInfo
+var typeInfoMapPopulate sync.Mutex // just for adding to typeInfoMap
+
+func getTypeInfo(t reflect.Type) *typeInfo {
+	if f, ok := typeInfoMap.Load(t); ok {
+		return f.(*typeInfo)
 	}
+	typeInfoMapPopulate.Lock()
+	defer typeInfoMapPopulate.Unlock()
+	return getTypeInfoLocked(t)
+}
 
-	w := h.bw
+func getTypeInfoLocked(t reflect.Type) *typeInfo {
+	if f, ok := typeInfoMap.Load(t); ok {
+		return f.(*typeInfo)
+	}
+	ti := &typeInfo{
+		rtype:       t,
+		isRecursive: typeIsRecursive(t),
+	}
+	typeInfoMap.Store(t, ti)
 
-	if v.CanInterface() {
-		// Use AppendTo methods, if available and cheap.
-		if v.CanAddr() && v.Type().Implements(appenderToType) {
-			a := v.Addr().Interface().(appenderTo)
-			size := h.scratch[:8]
-			record := a.AppendTo(size)
-			binary.LittleEndian.PutUint64(record, uint64(len(record)-len(size)))
-			w.Write(record)
-			return
+	switch t.Kind() {
+	case reflect.Map:
+		ti.keyTypeInfo = getTypeInfoLocked(t.Key())
+		fallthrough
+	case reflect.Ptr, reflect.Slice, reflect.Array:
+		ti.eleTypeInfo = getTypeInfoLocked(t.Elem())
+	}
+	return ti
+}
+
+// typeIsRecursive reports whether t has a path back to itself.
+//
+// For interfaces, it currently always reports true.
+func typeIsRecursive(t reflect.Type) bool {
+	inStack := map[reflect.Type]bool{}
+
+	var stack []reflect.Type
+
+	var visitType func(t reflect.Type) (isRecursiveSoFar bool)
+	visitType = func(t reflect.Type) (isRecursiveSoFar bool) {
+		switch t.Kind() {
+		case reflect.Bool,
+			reflect.Int,
+			reflect.Int8,
+			reflect.Int16,
+			reflect.Int32,
+			reflect.Int64,
+			reflect.Uint,
+			reflect.Uint8,
+			reflect.Uint16,
+			reflect.Uint32,
+			reflect.Uint64,
+			reflect.Uintptr,
+			reflect.Float32,
+			reflect.Float64,
+			reflect.Complex64,
+			reflect.Complex128,
+			reflect.String,
+			reflect.Func:
+			return false
+		}
+		if t.Size() == 0 {
+			return false
+		}
+		if inStack[t] {
+			return true
 		}
+		stack = append(stack, t)
+		inStack[t] = true
+		defer func() {
+			delete(inStack, t)
+			stack = stack[:len(stack)-1]
+		}()
+
+		switch t.Kind() {
+		default:
+			panic("unhandled kind " + t.Kind().String())
+		case reflect.Interface:
+			// Assume the worst for now. TODO(bradfitz): in some cases
+			// we should be able to prove that it's not recursive. Not worth
+			// it for now.
+			return true
+		case reflect.Array, reflect.Chan, reflect.Pointer, reflect.Slice:
```

Comment:
Do we really need to visit a Chan?

Reply:
Yes:

```
	type RecursiveChan chan *RecursiveChan
```

I added two new tests:

```
		{val: make(RecursiveChan), want: true},
		{val: make(chan int), want: false},
```

Reply:
Admittedly we won't be able to hash them anyway, but I'd rather this func be as correct as possible.

Reply:
I guess there is technically a cycle in the type-graph, but not a cycle in the value-graph (for hashing concerns) since we will never try to hash a channel.

SGTM.